xds/src/test/java/io/grpc/xds/XdsClientImplTest.java

```diff
@@ -0,0 +1,1373 @@
+import org.mockito.MockitoAnnotations;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+/**
+ * Tests for {@link XdsClientImpl}.
+ */
+public class XdsClientImplTest {
+
+  private static final String HOSTNAME = "foo.googleapis.com";
+  private static final int PORT = 8080;
+
+  private static final Node NODE = Node.getDefaultInstance();
+  private static final FakeClock.TaskFilter RPC_RETRY_TASK_FILTER =
+      new FakeClock.TaskFilter() {
+        @Override
+        public boolean shouldAccept(Runnable command) {
+          return command.toString().contains(XdsClientImpl.RpcRetryTask.class.getSimpleName());
+        }
+      };
+
+  @Rule
+  public final GrpcCleanupRule cleanupRule = new GrpcCleanupRule();
+
+  private final SynchronizationContext syncContext = new SynchronizationContext(
+      new Thread.UncaughtExceptionHandler() {
+        @Override
+        public void uncaughtException(Thread t, Throwable e) {
+          throw new AssertionError(e);
+        }
+      });
+  private final FakeClock fakeClock = new FakeClock();
+
+  private final Queue<StreamObserver<DiscoveryResponse>> responseObservers = new ArrayDeque<>();
+  private final Queue<StreamObserver<DiscoveryRequest>> requestObservers = new ArrayDeque<>();
+
+  @Mock
+  private AggregatedDiscoveryServiceImplBase mockedDiscoveryService;
+  @Mock
+  private BackoffPolicy.Provider backoffPolicyProvider;
+  @Mock
+  private BackoffPolicy backoffPolicy1;
+  @Mock
+  private BackoffPolicy backoffPolicy2;
+  @Mock
+  private ConfigWatcher configWatcher;
+
+  private ManagedChannel channel;
+  private XdsClientImpl xdsClient;
+
+  @Before
+  public void setUp() throws IOException {
+    MockitoAnnotations.initMocks(this);
+    when(backoffPolicyProvider.get()).thenReturn(backoffPolicy1, backoffPolicy2);
+    when(backoffPolicy1.nextBackoffNanos()).thenReturn(10L, 100L);
+    when(backoffPolicy2.nextBackoffNanos()).thenReturn(20L, 200L);
+
+    String serverName = InProcessServerBuilder.generateName();
+    AggregatedDiscoveryServiceImplBase serviceImpl = new AggregatedDiscoveryServiceImplBase() {
+      @Override
+      public StreamObserver<DiscoveryRequest> streamAggregatedResources(
+          final StreamObserver<DiscoveryResponse> responseObserver) {
+        responseObservers.offer(responseObserver);
+        @SuppressWarnings("unchecked")
+        StreamObserver<DiscoveryRequest> requestObserver = mock(StreamObserver.class);
+        requestObservers.offer(requestObserver);
+        Answer<Void> closeRpc = new Answer<Void>() {
+          @Override
+          public Void answer(InvocationOnMock invocation) {
+            responseObserver.onCompleted();
+            return null;
+          }
+        };
+        doAnswer(closeRpc).when(requestObserver).onCompleted();
+
+        return requestObserver;
+      }
+    };
+    mockedDiscoveryService =
+        mock(AggregatedDiscoveryServiceImplBase.class, delegatesTo(serviceImpl));
```

Comment:
This mock is redundant. All you need is consistently using `responseObservers.poll()` which should be the single source of truth. Now `mockedDiscoveryService` is doing the same thing.

Reply:
I want an __explicit__ evidence of each individual RPC call, which is crucial for doing step-by-step checks in retry tests. `responseObservers.poll()` doesn't give you direct proof of individual RPC call happening.

Reply:
Okay, I see you use `inOrder` for `mockedDiscoveryService`. That's not possible for `responseObservers.poll()`.